core/src/main/c/share/txn_board.cpp

```diff
@@ -25,127 +25,157 @@
 #include <cstdint>
 #include "util.h"
 #include "jni.h"
+#include <atomic>
+#include <algorithm>
 
 #define COUNTER_T uint16_t
 
-template<typename T>
-void set_max_atomic(T *slot, T value) {
-    do {
-        T current = __atomic_load_n(slot, __ATOMIC_RELAXED);
-        if (value <= current || __sync_val_compare_and_swap(slot, current, value) == current) {
-            break;
-        }
-    } while (true);
-}
-
-template<typename NEXT, typename T>
-inline T atomic_next(volatile T *val, NEXT next) {
-    do {
-        T current = __atomic_load_n(val, __ATOMIC_RELAXED);
-        T n = next(current);
-        if (__sync_val_compare_and_swap(val, current, n) == current) {
-            return n;
-        }
-    } while (true);
-}
-
 template<typename T>
 class txn_scoreboard_t {
     uint32_t mask = 0;
     uint32_t size = 0;
-    int64_t max = 0;
-    // 1-based min txn that is in-use
-    // we have to use 1 based txn to rule out possibility of 0 txn
-    // 0 is initial value when shared memory is created
-    int64_t min = L_MAX;
-    T counts[];
-
-    inline static T inc(T val) {
-        return val + 1;
+    std::atomic<int64_t> max{0};
+    std::atomic<int64_t> min{0};
+    std::atomic<T> counts[];
+
+    template<typename TT>
+    inline static TT set_max_atomic(std::atomic<TT> &slot, TT value) {
+        TT current = slot.load();
+        while (value > current && !slot.compare_exchange_weak(current, value));
+        return std::max(value, current);
     }
 
-    inline static T dec(T val) {
-        return val - 1;
+    inline std::atomic<T> &get_counter(const int64_t offset) {
+        return counts[offset & mask];
     }
 
-public:
-
-    inline int64_t get_min() {
-        return __atomic_load_n(&min, __ATOMIC_RELAXED);
-    }
+    inline bool increment_count(int64_t txn) {
+        // Increment txn count
+        // but do not allow to use txn below max value
+        // Once there is count for txn 100
+        // there cannot be increments for txn 99, 98 etc
+        // When an attempt to acquire below max happens
+        // the attempt is rejected and TableReader should re-read _txn file
+        auto current_max = max.load();
+        if (txn < current_max || txn - min.load() >= size) {
+            return false;
+        }
+        get_counter(txn)++; // atomic
 
-    inline int64_t get_max() {
-        return __atomic_load_n(&max, __ATOMIC_RELAXED);
-    }
+        current_max = max.load(std::memory_order_acquire);
```

Comment:
nit: since we intentionally avoid non-default memory orders, we probably should go with the default one here:
```suggestion
        current_max = max.load();
```